Lib/tempfile.py

```diff
@@ -518,18 +530,32 @@ def __iter__(self):
         for line in self.file:
             yield line
 
+    def __del__(self):
```

Comment:
I'd add the class attributes `file = None` and `_cleanup_called = False`, and a common `_cleanup()` method. For example:

```python
    file = None
    _cleanup_called = False

    def _cleanup(self):
        if self._cleanup_called or self.file is None:
            return
        self._cleanup_called = True
        self.close()
        # If the file is to be deleted, but not on close, delete it now.
        if self.delete and not self.delete_on_close:
            try:
                _os.unlink(self.name)
            # It is okay to ignore FileNotFoundError. The file may have
            # been deleted already.
            except FileNotFoundError:
                pass

    def __exit__(self, exc, value, tb):
        result = self.file.__exit__(exc, value, tb)
        self._cleanup()
        return result

    def __del__(self):
        self._cleanup()
```

Reply:
@eryksun , I am sorry if this is a stupid question, but can you please explain me why you use **class attributes** `name = None` and `_cleanup_called = False` ? Why can't they be an instance attributes:

```py
def __init__(self, .....)
    name = None
    _cleanup_called = False
```

Class attributes are normally used to represent data shared across all instances of the class. Surely we want to have these flags to be specific for every temporary file instance (or am I badly missing something?)

And then in this line
```py
    self._cleanup_called = True
```
you create already an instance attribute with the same name. So, what is the point?

I noticed, that the same trick is used in the `_TemporaryFileCloser`

```py
class _TemporaryFileCloser:
    """A separate object allowing proper closing of a temporary file's
    underlying file object, without adding a __del__ method to the
    temporary file."""

    file = None  # Set here since __del__ checks it
    close_called = False
```

Reply:
> can you please explain me why you use class attributes

I copied the pattern from `_TemporaryFileCloser`. I guess in some cases it could be safer than relying only on the instance attribute. For example, if `self._closer` were assigned first in `__init__()`, and instantiating the closer failed, then the wrapper's `__del__()` finalizer would see that `self.name is None` (the default value set as a class attribute) and thus skip trying to delete the file. That leaves it up to `NamedTemporaryFile()` to delete the file, which to me seems the proper thing to do since instantiating the wrapper failed.

Also, in the case of attributes such as `close_called`, which usually have the default value for the normal lifetime of an instance, using a class attribute can reduce the memory footprint. This could be a significant benefit if there are thousands of instances, but generally it's not a concern.

Reply:
Speaking of `NamedTemporaryFile()`, handling the case where creating the wrapper fails has to be updated for `delete_on_close` as well. For example:

```python
    except:
        if name is not None and not (
          _os.name == 'nt' and delete and delete_on_close):
            _os.unlink(name)
        raise
```